vowpalwabbit/namespaced_features.cc

```diff
@@ -0,0 +1,200 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include "namespaced_features.h"
+#include "vw_exception.h"
+
+using namespace VW;
+
+features* namespaced_features::get_feature_group(uint64_t hash)
+{
+  auto it = _hash_to_index_mapping.find(hash);
+  if (it == _hash_to_index_mapping.end()) { return nullptr; }
+
+  return &_feature_groups[it->second];
+}
+
+const features* namespaced_features::get_feature_group(uint64_t hash) const
+{
+  auto it = _hash_to_index_mapping.find(hash);
+  if (it == _hash_to_index_mapping.end()) { return nullptr; }
+
+  return &_feature_groups[it->second];
+}
+
+std::vector<namespace_index> namespaced_features::get_indices() const
+{
+  auto indices_copy = _namespace_indices;
+  std::sort(indices_copy.begin(), indices_copy.end());
+  auto last = std::unique(indices_copy.begin(), indices_copy.end());
+  indices_copy.erase(last, indices_copy.end());
+  return indices_copy;
+}
+
+std::pair<namespaced_features::indexed_iterator, namespaced_features::indexed_iterator>
+namespaced_features::get_namespace_index_groups(namespace_index ns_index)
+{
+  return std::make_pair(namespace_index_begin(ns_index), namespace_index_end(ns_index));
+}
+
+std::pair<namespaced_features::const_indexed_iterator, namespaced_features::const_indexed_iterator>
+namespaced_features::get_namespace_index_groups(namespace_index ns_index) const
+{
+  return std::make_pair(namespace_index_begin(ns_index), namespace_index_end(ns_index));
+}
+
+features& namespaced_features::get_or_create_feature_group(uint64_t hash, namespace_index ns_index)
```

Comment:
Correct, but it is needed if you wish to actually create the feature group. Unfortunately since there were *always* 256 of these the majority of call sites expect it to already exist